Modified CheckFingerprint to be able to take a VCF as INPUT

gbggrant opened the pull request:
Modified CheckFingerprint to be able to take a VCF as INPUT (as opposed to a SAM/BAM).
Added specialized method in FingerprintChecker for that purpose
Generalized certain methods to handle VCF vs BAM/SAM as input.

yfarjoun commented on src/main/java/picard/cmdline/programgroups/Fingerprinting.java:
```diff
@@ -0,0 +1,12 @@
+
+import picard.cmdline.CommandLineProgramGroup;
+/**
+ * @author nhomer
```
@nhomer?

gbggrant replied:
no mhre

yfarjoun commented on src/main/java/picard/cmdline/programgroups/Fingerprinting.java:
```diff
@@ -0,0 +1,12 @@
+package picard.cmdline.programgroups;
```
do we normally license these files?

gbggrant replied:
Evidently not, but I'll add the license

yfarjoun commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -33,48 +34,51 @@
             "which represents the relative likelihood of the sequence data originating from the same " +
             "sample as the genotypes vs. from a random sample.  Two outputs are produced: (1) a summary " +
-            "metrics file that gives metrics related single read group (lane or index within a lane) " +
+            "metrics file that gives metrics at the single sample level (if the input was a VCF) or at the read " +
```
vcf's can have more than one sample. please inform here what happens when N!=1 (both 0 and >1)

gbggrant replied:
done.

yfarjoun commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -33,48 +34,51 @@
             "and the detail metrics having a file extension '" + CheckFingerprint.FINGERPRINT_DETAIL_FILE_SUFFIX + "'.";
 
-    @Option(shortName=StandardOptionDefinitions.INPUT_SHORT_NAME, doc = "Input SAM or BAM file.")
+    @Option(shortName=StandardOptionDefinitions.INPUT_SHORT_NAME, doc = "Input file SAM/BAM or VCF.  (Single Sample)")
```
would it be difficult to make it work for a multi-sample vcf while you're at it?

yfarjoun replied:
I see that there are several places where a single EXPECTED_SAMPLE_ALIAS is assumed...still, in the case that a SAMPLE_ALIAS is provided, it could be used for both the  INPUT and the GENOTYPES files.

gbggrant replied:
Yes.  But I envision a case where the sample alias in the input VCF is not the same as that in the genotypes file, so I will supply an OBSERVED_SAMPLE_ALIAS parameter.

yfarjoun commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -89,10 +93,10 @@
-            "genotypes to use from the VCF file.  If not supplied the sample name from the BAM read group header " +
-            "is used instead.")
-    public String SAMPLE_ALIAS;
+    @Option(optional=true, doc = "This parameter can be used to specify which sample's genotypes to use from the " +
```
you are breaking backwards compatibility, could you make SAMPLE_ALIAS the short name for the argument?

gbggrant replied:
Yeah - that's a really good point and suggestion (I was considering reverting the name).

yfarjoun commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -129,37 +134,65 @@ protected int doWork() {
 
-        // If sample alias isn't supplied, assume it's the one from the INPUT file's RGs
-        if (SAMPLE_ALIAS == null) {
+            // Verify that there's only one sample in the SAM/BAM.
```
this verification should only happen if EXPECTED_SAMPLE_ALIAS is null, right?

yfarjoun replied:
the if statement below needs to be moved up.

gbggrant replied:
No.  I still want to validate that there's only one sample in the bam, so that's done in the for loop starting at line 154.  If there's more than one sample in the BAM, then an exception will be thrown, otherwise, observedSampleAlias will be defined.

The if statement at 165 checks if the user set the EXPECTED_SAMPLE_ALIAS to null and if so sets it to the observedSampleAlias I set as a result of the for loop above.

yfarjoun replied:
got it. so it's the same result, but in a different way. OK.

yfarjoun commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -129,37 +134,65 @@ protected int doWork() {
+            final VCFFileReader fileReader = new VCFFileReader(INPUT, false);
+            final VCFHeader fileHeader = fileReader.getFileHeader();
+            if (fileHeader.getNGenotypeSamples() != 1) {
+                throw new PicardException("INPUT VCF file must not contain data from multiple samples.");
```
a vcf could also have getNGenotypeSamples()==0...so your error message will be incorrect.

gbggrant replied:
True, so true.

yfarjoun commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -217,12 +252,25 @@ else if (!SAMPLE_ALIAS.equals(rec.getSample())) {
     }
+
+    protected String[] customCommandLineValidation() {
+        IOUtil.assertFileIsReadable(INPUT);
```
oh, I see, you are validating Readability here...

yfarjoun replied:
maybe move all the IO Asserts here for consistency?

yfarjoun replied:
nevermind. just put a comment that INPUT asserted to be Readable in  customCommandLineValidation()

gbggrant replied:
I put a copy back where it was.  Needed to validate readability here since this is done before doWork is hit.

yfarjoun commented on src/main/java/picard/fingerprint/CrosscheckReadGroupFingerprints.java:
```diff
@@ -50,7 +50,7 @@
 @CommandLineProgramProperties(
         usage = "Checks if all read groups within a set of BAM files appear to come from the same individual",
```
actually, this seems to be incorrect: "a set of..." program only takes a single file as input.

gbggrant replied:
Hmm.  Not my code, but looks like it takes a list of bams as an input and uses them

yfarjoun commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -527,7 +526,7 @@ public IntervalList getLociToGenotype(final Collection<Fingerprint> fingerprints
                 fingerprintsByReadGroup.values().forEach(combinedFp::merge);
 
-                final FingerprintResults results = new FingerprintResults(f, specificSample);
+                final FingerprintResults results = new FingerprintResults(f, specificSample, specificSample);
```
it's a little odd to have both the readgroup and the sample be the sample_alias. wouldn't it be better to have the readgroup be null or the empty string in this case?

gbggrant replied:
Yeah, I agree.  That's what the original code was doing, so I left it that way.  I like your idea of changing it to null

yfarjoun commented on src/main/java/picard/fingerprint/FingerprintResults.java:
```diff
@@ -29,27 +29,31 @@
     }
 
-    public File getSamFile() { return samFile; }
+    public File getInputFile() { return inputFile; }
```
how is this used? what is this for?

yfarjoun replied:
nm

yfarjoun commented on src/main/java/picard/fingerprint/FingerprintResults.java:
```diff
@@ -29,27 +29,31 @@
-    public FingerprintResults(final File samFile, final String readGroup) {
-        this.samFile = samFile;
+    public FingerprintResults(final File inputFile, final String readGroup, final String sampleAlias) {
+        this.inputFile = inputFile;
```
why do you needs to know the inputFile here?

yfarjoun replied:
nevermind. I see. you were just changing the samFile to inputFile...OK.

yfarjoun commented:
Back to you @gbggrant. 

comments:
0. what does this addition do that crossCheckGenotypes doesn't? is it the creation of a fingerprint metric? Does this supersede CrossCheckGenotypes? should it be deprecated now?
1. A test or two would be good. I can help you come up with some case examples, and a haplotype_database to use
2. other comments in-line

gbggrant commented on src/main/java/picard/fingerprint/CheckFingerprint.java:
```diff
@@ -129,37 +134,65 @@ protected int doWork() {
             outputSummaryMetricsFile = new File(OUTPUT + FINGERPRINT_SUMMARY_FILE_SUFFIX);
         }
 
-        IOUtil.assertFileIsReadable(INPUT);
```
I think I'll put it back even though it is checked in CustomCommandLineValidation - it's a pretty quick check.

gbggrant commented:
Thanks @yfarjoun I think I have addressed all of your comments. 
0.  This mostly is useful for the arrays workflow where we have an input single-sample VCF (the output of the autocall) and we want to check that against the fingerprint vcf AND generate the metrics file. 
0a.  I don't think that CrossCheckGenotypes can be deprecated because it does an all-pairs comparison and spits out best matches.
1.  I would appreciate some help on examples - particularly on how to handle what haplotype db to use.

gbggrant commented:
@yfarjoun I have added unit tests for this.  Could you please take another look?

yfarjoun commented on src/test/java/picard/fingerprint/FingerprintCheckerTest.java:
```diff
@@ -19,6 +21,11 @@
+    private static final double DELTA = 1e-6;
+
+    private static final File TEST_DATA_DIR = new File("testdata/picard/fingerprint/");
+    private static final File TEST_HAPLOTYPE_DATABASE_MAP = new File(TEST_DATA_DIR, "Homo_sapiens_assembly19.haplotype_database.txt");
```
can you change the name to make it clear that this is a small DB?

gbggrant commented:
@yfarjoun Have changed vcfs to be uncompressed.  Changed name of Haplotype database file.  Also reverted an earlier change of mine (the addition of the column 'OBSERVED_SAMPLE_ALIAS' to the summary metrics file which I didn't think was needed in retrospect.

yfarjoun commented on testdata/picard/fingerprint/NA12892.fp.vcf:
```diff
@@ -0,0 +1,102 @@
+##contig=<ID=NC_007605,length=171823>
+##fileDate=Wed Sep 14 12:02:21 EDT 2016
+##reference=file:///seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.fasta
+##source=Genotypes downloaded from gap.broadinstitute.org by DownloadGenotypes.jar
```
redact?

yfarjoun commented on testdata/picard/fingerprint/NA12891.vcf:
```diff
@@ -0,0 +1,102 @@
+##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency, for each ALT allele, in the same order as listed">
+##INFO=<ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes">
+##INFO=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth; some reads may have been filtered">
+##Illumina manifest file name=PsychChip_v1-1_15073391_A1.csv
```
needs redacting?

yfarjoun commented on testdata/picard/fingerprint/NA12891.vcf:
```diff
@@ -0,0 +1,102 @@
+##fileformat=VCFv4.1
```
really 4.1?

gbggrant replied:
Really

gbggrant commented:
Okay @yfarjoun Have redacted information from vcf files.
- FingerprintChecker wants indexed vcfs, and I don't know an easy way to index on the fly, so have left the index files checked in.
- VCF format remains 4.1 - we are creating another ticket to see why this is set and if it is appropriate or can be revved.

yfarjoun commented:
I think you need to re-write the vcf and enable indexing on the fly for this.

yfarjoun commented:
you changed the vcfs, did you re-index them?

gbggrant commented:
Yes, I did re-index them (the binaries were updated).
Will look into creating the index on the fly for the test.

gbggrant commented:
@yfarjoun have removed the index files and created a utility class to create indexed vcfs on the fly and clean up after.

gbggrant commented:
Thanks @yfarjoun
